lib/spack/spack/test/spec_dag.py

```diff
@@ -989,3 +1021,99 @@ def test_spec_tree_respect_deptypes(self):
+    # dependencies from the same package
+    root = Spec('b').concretized()
+    link_run_spec = Spec('c@1.0').concretized()
+    build_spec = Spec('c@2.0').concretized()
+
+    root.add_dependency_edge(link_run_spec, deptype='link')
+    root.add_dependency_edge(link_run_spec, deptype='run')
+    root.add_dependency_edge(build_spec, deptype='build')
+
+    # Check dependencies from the perspective of root
+    assert len(root.dependencies()) == 2
+    assert all(x.name == 'c' for x in root.dependencies())
+
+    assert '@2.0' in root.dependencies(name='c', deptype='build')[0]
+    assert '@1.0' in root.dependencies(name='c', deptype=('link', 'run'))[0]
+
+    # Check parent from the perspective of the dependencies
+    assert len(build_spec.dependents()) == 1
+    assert len(link_run_spec.dependents()) == 1
+    assert build_spec.dependents() == link_run_spec.dependents()
+    assert build_spec != link_run_spec
+
+
+def test_synthetic_construction_bootstrapping(mock_packages, config):
+    # Construct the following spec:
+    #
+    #  b@2.0
+    #    | build
+    #  b@1.0
+    #
+    root = Spec('b@2.0').concretized()
+    bootstrap = Spec('b@1.0').concretized()
+
+    root.add_dependency_edge(bootstrap, deptype='build')
+
+    assert len(root.dependencies()) == 1
+    assert root.dependencies()[0].name == 'b'
+    assert root.name == 'b'
+
+
+def test_addition_of_different_deptypes_in_multiple_calls(mock_packages, config):
+    # Construct the following spec:
+    #
+    #  b@2.0
+    #    | build,link,run
+    #  b@1.0
+    #
+    # with three calls and check we always have a single edge
+    root = Spec('b@2.0').concretized()
+    bootstrap = Spec('b@1.0').concretized()
+
+    for current_deptype in ('build', 'link', 'run'):
+        root.add_dependency_edge(bootstrap, deptype=current_deptype)
+
+        # Check edges in dependencies
+        assert len(root.edges_to_dependencies()) == 1
+        forward_edge = root.edges_to_dependencies(deptype=current_deptype)[0]
+        assert current_deptype in forward_edge.deptypes
+        assert id(forward_edge.parent) == id(root)
+        assert id(forward_edge.spec) == id(bootstrap)
+
+        # Check edges from dependents
+        assert len(bootstrap.edges_from_dependents()) == 1
+        backward_edge = bootstrap.edges_from_dependents(deptype=current_deptype)[0]
+        assert current_deptype in backward_edge.deptypes
+        assert id(backward_edge.parent) == id(root)
+        assert id(backward_edge.spec) == id(bootstrap)
+
+
+@pytest.mark.parametrize('c1_deptypes,c2_deptypes', [
+    ('link', ('build', 'link')),
+    (('link', 'run'), ('build', 'link'))
+])
+def test_adding_same_deptype_with_the_same_name_raises(
+        mock_packages, config, c1_deptypes, c2_deptypes
+):
+    p = Spec('b@2.0').concretized()
+    c1 = Spec('b@1.0').concretized()
+    c2 = Spec('b@2.0').concretized()
+
+    p.add_dependency_edge(c1, deptype=c1_deptypes)
+    with pytest.raises(spack.error.SpackError):
```

Comment:
I verified that this raises the right exception but for the future: this test should probably not have the parent *and* the dependencies as `b`.  If we added cycle checking or something I could see this error getting confused with another.